packages/controllers/src/permissions/Caveat.ts

```diff
@@ -0,0 +1,254 @@
+
+/**
+ * A function for applying caveats to a restricted method request.
+ *
+ * @template ParentCaveat - The caveat type associated with this decorator.
+ * @param decorated - The restricted method implementation to be decorated.
+ * The method may have already been decorated with other caveats.
+ * @param caveat - The caveat object.
+ * @returns The decorate restricted method implementation.
+ */
+export type CaveatDecorator<ParentCaveat extends CaveatConstraint> = (
+  decorated: AsyncRestrictedMethod<RestrictedMethodParameters, Json>,
+  caveat: ParentCaveat,
+) => AsyncRestrictedMethod<RestrictedMethodParameters, Json>;
+
+/**
+ * TODO:docs
+ */
+type ExtractCaveatValueFromDecorator<Decorator extends CaveatDecorator<any>> =
+  Decorator extends (
+    decorated: any,
+    caveat: infer ParentCaveat,
+  ) => AsyncRestrictedMethod<any, any>
+    ? ParentCaveat extends CaveatConstraint
+      ? ParentCaveat['value']
+      : never
+    : never;
+
+/**
+ * TODO:docs
+ */
+export type CaveatValidator<ParentCaveat extends CaveatConstraint> = (
+  caveat: { type: ParentCaveat['type']; value: unknown },
+  origin?: string,
+  target?: string,
+) => void;
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationConstraint = {
+  /**
+   * The string type of the caveat.
+   */
+  type: string;
+
+  /**
+   * The decorator function used to apply the caveat to restricted method
+   * requests.
+   */
+  decorator: CaveatDecorator<any>;
+
+  /**
+   * The validator function used to validate caveats of the associated type
+   * whenever they are instantiated. Caveat are instantiated whenever they are
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<any>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecification<Type extends string> = {
+  /**
+   * The string type of the caveat.
+   */
+  type: Type;
+
+  /**
+   * The decorator function used to apply the caveat to restricted method
+   * requests.
+   */
+  decorator: CaveatDecorator<any>;
+
+  /**
+   * The validator function used to validate caveats of the associated type
+   * whenever they are instantiated. Caveat are instantiated whenever they are
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<any>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationMap<
+  Specification extends CaveatSpecification<string>,
```

Comment:
This should be using `CaveatSpecificationConstraint` I think? Since we don't know the key at this point.

```suggestion
  Specification extends CaveatSpecificationConstraint,
```